blackjax/inference/base.py

```diff
@@ -0,0 +1,200 @@
+
+    This implementation is very general and should accomodate most variations
+    on the method.
+
+    Parameters
+    ----------
+    proposal_generator:
+        The function used to propose a new state for the chain. For vanilla HMC this
+        function integrates the trajectory over many steps, but gets more involved
+        with other algorithms such as empirical and dynamical HMC.
+    momentum_generator:
+        A function that returns a new value for the momentum when called.
+    kinetic_energy:
+        A function that computes the current state's kinetic energy.
+    potential:
+        The potential function that is being explored, equal to minus the likelihood.
+    divergence_threshold:
+        The maximum difference in energy between the initial and final state
+        after which we consider the transition to be divergent.
+
+    Returns
+    -------
+    A kernel that moves the chain by one step when called.
+
+    References
+    ----------
+    .. [1]: Duane, Simon, et al. "Hybrid monte carlo." Physics letters B
+            195.2 (1987): 216-222.
+    .. [2]: Neal, Radford M. "An improved acceptance procedure for the
+            hybrid Monte Carlo algorithm." Journal of Computational Physics 111.1
+            (1994): 194-203.
+    .. [3]: Betancourt, Michael. "A conceptual introduction to
+            Hamiltonian Monte Carlo." arXiv preprint arXiv:1701.02434 (2018).
+    .. [4]: "Hamiltonian Mechanics", Wikipedia.
+            https://en.wikipedia.org/wiki/Hamiltonian_mechanics#Deriving_Hamilton's_equations
+    .. [5]: Betancourt, Michael, et al. "The geometric foundations
+            of hamiltonian monte carlo." Bernoulli 23.4A (2017): 2257-2298.
+
+    """
+
+    def kernel(
+        rng_key: jax.random.PRNGKey, state: HMCState
+    ) -> Tuple[HMCState, HMCInfo]:
+        """Moves the chain by one step using the Hamiltonian dynamics.
+
+        Parameters
+        ----------
+        rng_key:
+           The pseudo-random number generator key used to generate random numbers.
+        state:
+            The current state of the chain: position, log-probability and gradient
+            of the log-probability.
+
+        Returns
+        -------
+        The next state of the chain and additional information about the current step.
+        """
+        key_momentum, key_integrator, key_accept = jax.random.split(rng_key, 3)
+
+        position, potential_energy, potential_energy_grad = state
+        momentum = momentum_generator(key_momentum, position)
+        energy = potential_energy + kinetic_energy(momentum, position)
+
+        proposal, proposal_info = proposal_generator(
+            key_integrator, HMCProposalState(position, momentum, potential_energy_grad)
+        )
+        new_position, new_momentum, new_potential_energy_grad = proposal
+
+        flipped_momentum = jax.tree_util.tree_multimap(lambda m: -1.0 * m, new_momentum)
+        new_potential_energy = potential_fn(new_position)
```

Comment:
We can avoid this PE evaluation by caching `potential_energy` or `log_prob` in `HMCProposalState`.

Reply:
Yes, you and @fehiepsi are right, I will edit the integrator.